### PR TITLE
GH-14846: [Dev] Support GitHub Releases in download_rc_binaries.py

### LINKED ISSUE #14846: [Dev] Update download_rc_binaries to be able to fetch from GitHub Releases
### Describe the enhancement requested

ADBC is using GitHub Releases instead of Artifactory; it would be nice to share a little bit of this infrastructure instead of having to replicate it all. (And eventually Arrow may be able to publish binaries on GitHub as well; we already do with Crossbow.)

See https://github.com/apache/arrow-adbc/pull/215#discussion_r1040123283

### Component(s)

Developer Tools

### PR DESCRIPTION
**Pull request by lidavidm**
* Closes: #14846

### CODE REVIEW

File: dev/release/download_rc_binaries.py [hunk 1]
```diff
@@ -125,6 +140,67 @@ class Maven(Downloader):
...
 
+class GitHub(Downloader):
+    def __init__(self, repository, tag):
+        super().__init__()
+        if repository is None:
+            raise ValueError("--repository is required")
+        if tag is None:
+            raise ValueError("--tag is required")
+        self.repository = repository
+        self.tag = tag
```
Comment (kou) [inline dev/release/download_rc_binaries.py:151]: Should we use `_` prefix here?

```suggestion
        self._repository = repository
        self._tag = tag
```

File: dev/release/download_rc_binaries.py [hunk 2]
```diff
@@ -125,6 +140,67 @@ class Maven(Downloader):
...
+            )
+            files.append((asset["name"], url))
+        return files
+
+    def _download_file(self, dest, asset):
+        name, url = asset
+
+        os.makedirs(dest, exist_ok=True)
+        dest_path = os.path.join(dest, name)
+        print("Downloading {} to {}".format(url, dest_path))
```
Comment (kou) [inline dev/release/download_rc_binaries.py:186]:
```suggestion
        print(f"Downloading {url} to {dest_path}")
```

File: dev/release/download_rc_binaries.py [hunk 3]
```diff
@@ -125,6 +140,67 @@ class Maven(Downloader):
...
+            return
+
+        return self._download_url(
+            url,
+            dest_path,
+            extra_args=[
+                "--header",
+                "Accept: application/octet-stream",
+                "--header",
+                "User-Agent: apache/arrow-adbc dev/release/download_release.py",
```
Comment (kou) [inline dev/release/download_rc_binaries.py:199]: Do we still need this?

**Review by kou — APPROVED** [review]
+1

File: dev/release/download_rc_binaries.py (hunk 3 above)
Comment (lidavidm) [reply dev/release/download_rc_binaries.py:199]: Hmm, probably not. It was for the API.

**Review by assignUser — APPROVED** [review]
Nice idea, LGTM! There is just a small lint issue:
```diff
@@ -128,7 +128,8 @@
                 os.remove(dest_path)
             except IOError:
                 pass
-            raise Exception(f"Downloading {url} failed\nstdout: {stdout}\nstderr: {stderr}")
+            raise Exception(
+                f"Downloading {url} failed\nstdout: {stdout}\nstderr: {stderr}")
 
 
 class Artifactory(Downloader):
```

**Pull request merged**